### DISCUSSION TITLE
How to prevent a block of text from splitting at page boundary

### DISCUSSION THREAD

**Discussion by ThinkTRG** [dolanmiu/docx#2037] [General]
I have some blocks of text wherein every line should be on the same page, without a page break splitting the block.

Here's an example:

```
Hdr1 Hdr2 Hdr3.....
This That The other
More Like The first
Just Keep Together!
```

What I **don't** want to happen is for a page break to split this block and force some of the lines onto a different page:

Hdr1 Hdr2 Hdr3.....
This That The other
-------------------  (page break)
More Like The first
Just Keep Together!
```

Sometimes this is called "widow/orphan control".

Is this something I can do with docx.js?

I searched a bit and found [this page](https://github.com/dolanmiu/docx/discussions/1832) but I don't think that's exactly what I'm looking for.

-- 
John Treankler

**Comment by ThinkTRG** [comment]
The [documentation for paragraphs](https://docx.js.org/#/usage/paragraph) references "Page Break Control" and `.keepLines()` and `.keepNext()`. There's [a link to an article](https://support.microsoft.com/en-us/office/change-the-line-spacing-in-word-04ada056-b8ef-4b84-87dd-5d7c28a85712?redirectSourcePath=%252fen-us%252farticle%252fKeep-lines-and-paragraphs-together-d72af534-926f-4c4b-830a-abfc2daa3bfa) on the help page, but the page at that link doesn't speak to widow/orphan control and I find no examples of how to code `keepLines()` or `keepNext()` in docx.js.

**Comment by ThinkTRG** [comment]
There's ["Widow Control" in the doc](https://docx.js.org/#/usage/paragraph?id=widow-control) which sounds exactly like what I want but I've hit the paragraph with both `widowControl: false,` and `widowControl: true,` with no effect.

**Comment by ThinkTRG** [comment]
Libre Office Writer app gives me what I want if I select the paragraph, then go to Format > Paragraph > Text Flow (tab) > ✅ Do not split paragraph.

**Reply by ThinkTRG** [reply]
So maybe what I'm asking is, how to set the 'do not split' attribute of a paragraph in docx.js?

**Reply by dolanmiu** [reply]
Can you try something like this:

```ts
const paragraph = new Paragraph({
    text: "Stay on the same page",
    keepLines: true,
    keepNext: true,
});
```

**Reply by ThinkTRG** [reply]
Thank you for that very quick response, @dolanmiu! I was certain I tried that combination earlier, but when I tried it just now, it seems to be working perfectly. I'll do some more testing and report my results back here.

**Reply by ThinkTRG** [reply]
Only the `keepLines` attribute of a Paragraph is needed to accomplish my goal. Here's how `keepLines` (left) and `keepNext` (right) in LibreOffice > Format > Paragrah > Text Flow: 
<img width="654" alt="KeepLines vs KeepNext" src="https://user-images.githubusercontent.com/114637599/227785182-87daa000-bbdd-4613-ac09-6a023e862119.png">
So `keepLines` implements `Do not split paragraph` which is exactly what I want. 😁😎👏🏼

**Reply by dolanmiu** [reply]
Good to know it worked! I will update the documentation later. Currently blocked by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64806

Documentation build is failing
